src/systems/tunnel-effect.js

```diff
@@ -0,0 +1,164 @@
+    this.isMoving = false;
+    this.isVR = false;
+    this.dt = 0;
+    this.isPostProcessingReady = false;
+    this.characterEl = document.querySelector(`a-entity[${this.data.targetComponent}]`);
+    if (this.characterEl) {
+      this._initPostProcessing = this._initPostProcessing.bind(this);
+      this.characterEl.addEventListener("componentinitialized", this._initPostProcessing);
+    } else {
+      console.warn("Could not find target component.");
+    }
+    this._enterVR = this._enterVR.bind(this);
+    this._exitVR = this._exitVR.bind(this);
+    this.scene.addEventListener("enter-vr", this._enterVR);
+    this.scene.addEventListener("exit-vr", this._exitVR);
+  },
+
+  pause: function() {
+    if (!this.characterEl) {
+      return;
+    }
+    this.characterEl.removeEventListener("componentinitialized", this._initPostProcessing);
+    this.scene.removeEventListener("enter-vr", this._enterVR);
+    this.scene.removeEventListener("exit-vr", this._exitVR);
+  },
+
+  play: function() {
+    this.scene.addEventListener("enter-vr", this._enterVR);
+    this.scene.addEventListener("exit-vr", this._exitVR);
+  },
+
+  tick: function(t, dt) {
+    this.dt = dt;
+
+    if (isDisabled || !this.isPostProcessingReady || !this.isVR) {
+      return;
+    }
+
+    const characterSpeed = this.characterComponent.velocity.length();
+    if (characterSpeed < CLAMP_SPEED) {
+      // the character stops, so we use the aframe default render func
+      const r = this.vignettePass.uniforms["radius"].value;
+      const softness = this.vignettePass.uniforms["softness"].value;
+      if (
+        this.isMoving &&
+        Math.abs(r - TARGET_RADIUS) > CLAMP_RADIUS &&
+        Math.abs(softness - TARGET_SOFTNESS) > CLAMP_SOFTNESS
+      ) {
+        if (this.deltaR !== 0 && this.deltaS !== 0) {
+          this.deltaR = TARGET_RADIUS - r;
+          this.deltaS = softness - TARGET_SOFTNESS;
+        }
+        const ratio = this.dt / FADE_TIMEOUT_MS;
+        this._updateVignettePass(r + this.deltaR * ratio, softness - this.deltaS * ratio, this.data.opacity);
+      } else {
+        this._exitTunnel();
+      }
+      return;
+    }
+
+    if (!this.isMoving) {
+      this.isMoving = true;
+      this._bindRenderFunc();
+    }
+    const { radius, minRadius, maxSpeed, softness, opacity } = this.data;
+    const r = (radius - minRadius) * (maxSpeed - characterSpeed) + minRadius;
+    this._updateVignettePass(r, softness, opacity);
+  },
+
+  _exitTunnel: function() {
+    this.scene.renderer.render = this.originalRenderFunc;
+    this.isMoving = false;
+  },
+
+  _initPostProcessing: function(event) {
+    if (event.detail.name === this.data.targetComponent) {
+      this.characterEl.removeEventListener("componentinitialized", this._initPostProcessing);
+      this.characterComponent = this.characterEl.components[this.data.targetComponent];
+      this._initComposer();
+    }
+  },
+
+  _enterVR: function() {
+    this.isVR = true;
```

Comment:
Unfortunately if you press the "f" key without a headset plugged in, the "entervr" event still fires. You go into full screen, but you're not actually in VR. This is not the biggest deal because if you press F11 you can go into fullscreen mode without going into VR, but at some point we may want to fix this.